src/main/java/org/opensearch/security/configuration/DlsQueryParser.java

```diff
@@ -15,151 +15,152 @@
-                dlsQueryBuilder.add(dlsQuery, Occur.SHOULD);
+            if (queryMapFunction != null) {
+                dlsQuery = queryMapFunction.apply(dlsQuery);
+            }
 
-                if (hasNestedMapping) {
-                    handleNested(queryShardContext, dlsQueryBuilder, dlsQuery);
-                }
+            dlsQueryBuilder.add(dlsQuery, Occur.SHOULD);
 
-            } catch (ExecutionException e) {
-                throw new IOException(e);
+            if (hasNestedMapping) {
+                handleNested(queryShardContext, dlsQueryBuilder, dlsQuery);
             }
         }
 
-        // no need for scoring here, so its possible to wrap this in a
-        // ConstantScoreQuery
-        return new ConstantScoreQuery(dlsQueryBuilder.build());
-
+        return dlsQueryBuilder;
+    }
+    
+    private static void handleNested(final QueryShardContext queryShardContext, final BooleanQuery.Builder dlsQueryBuilder,
+            final Query parentQuery) {
+        final BitSetProducer parentDocumentsFilter = queryShardContext.bitsetFilter(NON_NESTED_QUERY);
+        dlsQueryBuilder.add(new ToChildBlockJoinQuery(parentQuery, parentDocumentsFilter), Occur.SHOULD);
     }
 
-    static ParsedQuery parse(final Set<String> unparsedDlsQueries, ParsedQuery originalQuery, final QueryShardContext queryShardContext,
-                             final NamedXContentRegistry namedXContentRegistry) throws IOException {
-        if (unparsedDlsQueries == null || unparsedDlsQueries.isEmpty()) {
-            return null;
-        }
-
-        final boolean hasNestedMapping = queryShardContext.getMapperService().hasNested();
-
-        BooleanQuery.Builder dlsQueryBuilder = new BooleanQuery.Builder();
-        dlsQueryBuilder.setMinimumNumberShouldMatch(1);
-
-        for (final String unparsedDlsQuery : unparsedDlsQueries) {
-            try {
-
-                final QueryBuilder qb = queries.get(unparsedDlsQuery, new Callable<QueryBuilder>() {
+    public QueryBuilder parse(String unparsedDlsQuery) {
+        try {
+            final QueryBuilder qb = parsedQueryCache.get(unparsedDlsQuery, new Callable<QueryBuilder>() {
 
-                    @Override
-                    public QueryBuilder call() throws Exception {
-                        final XContentParser parser = JsonXContent.jsonXContent.createParser(namedXContentRegistry, THROW_UNSUPPORTED_OPERATION, unparsedDlsQuery);
-                        final QueryBuilder qb = AbstractQueryBuilder.parseInnerQueryBuilder(parser);
-                        return qb;
-                    }
+                @Override
+                public QueryBuilder call() throws Exception {
+                    final XContentParser parser = JsonXContent.jsonXContent.createParser(namedXContentRegistry,
+                            DeprecationHandler.THROW_UNSUPPORTED_OPERATION, unparsedDlsQuery);
+                    final QueryBuilder qb = AbstractQueryBuilder.parseInnerQueryBuilder(parser);
```

Comment:
unnecessary variable? can directly be `return`ed

Reply:
Agreed, I changed to code accordingly